torchao/prototype/dtypes/uint2.py

```diff
@@ -0,0 +1,446 @@
+        tl.store(output + offsets * 4 + 3, fourth_elements, mask=mask)
+
+    def unpack_uint8_to_trinary2(uint8_data: torch.Tensor) -> torch.Tensor:
+        uint8_data = uint8_data.to('cuda')
+        shape = uint8_data.shape
+        output = torch.empty(up_size(shape), dtype=torch.int8, device='cuda')
+        n_elements = uint8_data.numel()
+        grid = lambda meta: (triton.cdiv(n_elements, meta['BLOCK_SIZE']),)
+        triton_unpack_uint8_to_trinary2[grid](uint8_data, output, n_elements, BLOCK_SIZE=1024)
+        return output
+
+    @triton.jit
+    def triton_unpack_uint2(uint8_data, output, n_elements, BLOCK_SIZE: tl.constexpr):
+        offsets = tl.arange(0, BLOCK_SIZE)
+        mask = offsets < n_elements
+
+        uint8_value = tl.load(uint8_data + offsets, mask=mask)
+
+        first_elements = (uint8_value >> 6) & 0b11
+        second_elements = (uint8_value >> 4) & 0b11
+        third_elements = (uint8_value >> 2) & 0b11
+        fourth_elements = uint8_value & 0b11
+
+        tl.store(output + offsets * 4 + 0, first_elements, mask=mask)
+        tl.store(output + offsets * 4 + 1, second_elements, mask=mask)
+        tl.store(output + offsets * 4 + 2, third_elements, mask=mask)
+        tl.store(output + offsets * 4 + 3, fourth_elements, mask=mask)
+
+    def unpack_uint2(uint8_data: torch.Tensor) -> torch.Tensor:
+        uint8_data = uint8_data.to('cuda')
+        shape = uint8_data.shape
+        output = torch.empty(up_size(shape), dtype=torch.uint8, device='cuda')
+        n_elements = uint8_data.numel()
+        grid = lambda meta: (triton.cdiv(n_elements, meta['BLOCK_SIZE']),)
+        triton_unpack_uint2[grid](uint8_data, output, n_elements, BLOCK_SIZE=1024)
+        return output
+
+    @triton.jit
+    def triton_pack_uint2(uint8_data, output, n_elements, BLOCK_SIZE: tl.constexpr):
+        offsets = tl.arange(0, BLOCK_SIZE)
+        mask = offsets * 4 < n_elements
+
+        first_elements = tl.load(uint8_data + offsets * 4 + 0, mask=mask)
+        second_elements = tl.load(uint8_data + offsets * 4 + 1, mask=mask)
+        third_elements = tl.load(uint8_data + offsets * 4 + 2, mask=mask)
+        fourth_elements = tl.load(uint8_data + offsets * 4 + 3, mask=mask)
+
+        packed_data = (first_elements << 6) | (second_elements << 4) | (third_elements << 2) | fourth_elements
+
+        tl.store(output + offsets, packed_data, mask=mask)
+
+    def pack_uint2(uint8_data: torch.Tensor) -> torch.Tensor:
+        uint8_data = uint8_data.to('cuda')
+        shape = uint8_data.shape
+        assert shape[-1] % 4 == 0
+        n_elements = uint8_data.numel()
+        packed_shape = down_size(shape)
+        output = torch.empty(packed_shape, dtype=torch.uint8, device='cuda')
+        grid = lambda meta: (triton.cdiv(n_elements, meta['BLOCK_SIZE'] * 4),)
+        triton_pack_uint2[grid](uint8_data, output, n_elements, BLOCK_SIZE=1024)
+        return output
+
+else:
```

Comment:
are the kernels compilable on CPU? just wondering if these should ever get called today or they're mostly in so we can eventually revert the triton kernels?

Reply:
Yes, That's right. Will eventually revert the triton kernels with torch.compile.